pydantic/env_settings.py

```diff
@@ -217,7 +217,22 @@ def field_is_complex(self, field: ModelField) -> Tuple[bool, bool]:
 
         return True, allow_json_failure
 
-    def explode_env_vars(self, field: ModelField, env_vars: Mapping[str, Optional[str]]) -> Dict[str, Any]:
+    @staticmethod
+    def next_field(field: Union[ModelField, None], key: str) -> Union[ModelField, None]:
+        if not field:
+            return None
+
+        if field.sub_fields:
+            # no support for Unions of complex BaseSettings fields
+            return None
```

Comment:
Maybe we can combine these two conditions:
```suggestion
        if not field or field.sub_fields:
            # no support for Unions of complex BaseSettings fields
            return None
```